Various relation dialog fixes and improvements, refs #10682 #13218 #13425 #13499

mInnes-archives opened the pull request:
This is a small bug fix addressing 500 errors being thrown after attempting to save various types of records (eg. ISAAR, ISDF, Accession, etc.) containing incomplete relationships. For example, the related entity is unspecified so the relationship is incomplete:

![emptyRelatedEntity](https://user-images.githubusercontent.com/76928773/112741336-d8a95b80-8f52-11eb-803f-773c90d3ebc2.png)

To address this, an additional string validator has been added requiring that the relevant value be set. The incomplete relationship can still be staged as part of the form, like in the image above, but this relationship entry will now be ignored when saving. No error is thrown and all other form values and relationships will save as expected.

mInnes-archives commented:
Noticed I needed to rebase and make some minor syntax fixes. I may have caused Syntax checks to bug out with all the force-pushes somehow. :sweat:

jraddaoui commented:
Hi @mInnes-archives,

Thanks for reporting the issue and submitting a PR. I see you also rebased your branch after the latest changes in qa/2.x, which added the syntax checks I mentioned you in your previous PR. We just announced this change in our coding standards in the forum, where you can find some information:

https://groups.google.com/g/ica-atom-users/c/edNrTGACuv4/m/-pPIhMYoCgAJ

As you can see in the failing check details, it's only complaining about indentation, so it should be an easy fix.

jraddaoui commented:
About the actual issue and changes, I believe @fiver-watson had a question about it and I'd love to see some sort of notification about the relation not being saved, but I'm not sure how hard would it be in this multi-row component form and it's probably needed in other places. So it looks good to me.

mInnes-archives commented:
> As you can see in the failing check details, it's only complaining about indentation, so it should be an easy fix.

Okay, now that I'm reading the details correctly, I see that. I'll get the plug-in installed to avoid this in future.

> About the actual issue and changes, I believe @fiver-watson had a question about it and I'd love to see some sort of notification about the relation not being saved, but I'm not sure how hard would it be in this multi-row component form and it's probably needed in other places. So it looks good to me.

Agreed. that would be ideal. This is somewhat of a quick fix for now.

jraddaoui commented:
> Okay, now that I'm reading the details correctly, I see that. I'll get the plug-in install to avoid this in future.

I see, the details are not very clear. I've tried to improve this in https://github.com/artefactual/atom/pull/1278, using a shorter diff format.

fiver-watson commented:
Hi @mInnes-archives! Thanks again for your continued contributions to the project! 

I just wanted to clarify the behavior on this. If I'm understanding correctly: 

* Adding an incomplete relationship currently throws a 500 error, which means that any other data entered into the edit page risks being lost
* This proposed fix will make these relationships fail silently - they will be ignored, allowing the page to be saved, and no error or warning will be thrown
* As a consequence, however, the user has no indication that some of the data they entered (even though incomplete) has not been saved

Is that correct?

If so, then while I see the motivation for this (if you spend a lot of time working on a description, it's frustrating to lose all that work because of one bad relationship element), but I also worry about not letting the user know that some data has been dropped. 

AtoM does have a warning notification that can be displayed to authenticated users on view pages following a save - we use it for post-submission validation (AtoM's always had a very permissive approach - you can technically save a completely blank record and it will work). For example, you'll see one of the warnings if lower-level dates are incompatible with those at higher levels, or if one of the elements the related standard considers mandatory are not filled in (for example, leaving the identifier, title, or extent field blank). 

As Radda has mentioned, I think it would be ideal to use one of these to inform the user about what's happened. I don't know if you have the time or inclination to add this - if so, I'm hoping that @jraddaoui can provide an implementation example from elsewhere in the code as a reference, and we might make this a valuable addition to the next major release. 

If not, I'm still inclined for us to accept the PR - I agree that this is probably better than potentially losing a lot of other user data when throwing the error. In that case, I will file a new issue and we'll try to revisit this and add the warning ourselves in the future, time allowing. 

Thoughts?

jraddaoui commented:
Here an example of a flash alert ...

https://github.com/artefactual/atom/blob/qa/2.x/apps/qubit/modules/informationobject/actions/updatePublicationStatusAction.class.php#L50-L51

I believe "notice" is a yellow notification, "error" is red and "info" is blue.

mInnes-archives commented:
@fiver-watson 
> I don't know if you have the time or inclination to add this

I'd be happy to dig a bit deeper and find an alternative solution which could provide better feedback for users. I'm just doing this in my spare time and I'm really just learning things as I go, so I'll see what I can find out over the next week or so.

@jraddaoui 
>Here an example of a flash alert ...

Thanks. :)

mInnes-archives commented:
### Cause of 500 errors

After digging around I found out the cause of the original 500 errors. Attempting to add incomplete relationships to the database entails trying to add `null` values to `NOT NULL` columns. To explore a fully permissive approach to this problem I tested the effects of changing this constraint but found that this ended up causing way more problems. Therefore, stopping these data from being submitted remains the best approach.

### Augmented solution

To augment my original solution I've added additional validation and error messaging as part of the relevant YUI dialogs. This should add sufficient feedback for users and halt the creation of these relationships. This solution has been applied to the following dialogs thus far:

- ISAAR - Related authority records
- ISAAR - Related resources
- ISDF - Related authority records
- ISDF - Related resources

This solution has yet to be added to the following due to the bugs outlined in the next section:

- Accession records - Related donors
- ISDF - Related functions

### Additional issues

I've identified some additional issues in doing this work. Some I've already addressed, such as typos or alignment issues, but the following remain:

- Dialogs which allow for on-the-fly record creation for new related entities causes both the new record to be created (as desired) as well an additional 'Untitled' record. The new dialog validator also seems to cause some problems when dealing with these newly created entities. Affects the following:
  - 'Accession - Related Donors'
  - Actor events for archival descriptions, but note that the 500 error outlined above isn't relevant for this. 

- Clicking the edit icon for 'ISDF - Related functions' relationships ~~when the current record being edited is the object of the relationship~~ seems to result in a failure to pull up some field values in the dialog.

I'd like to try my hand at addressing these. so these fixes should be forthcoming.

### Images
![ErrorExample](https://user-images.githubusercontent.com/76928773/115157498-9ce84a00-a057-11eb-8ab9-740ba76477d7.png)
*Error example from ISAAR - Related resources*

jraddaoui commented:
Thanks for the update and fixes @mInnes-archives, it's looking great so far!

mInnes-archives commented:
Issues related to the 'ISDF - Related functions' dialog are resolved. The result returned from the Ajax response didn't accommodate the layout type used for ISDF records, so the name wasn't able to be retrieved. Having to rely on Ajax requests isn't great, but it works now.

I've also modified the title display formatting when viewing ISDF records to be more appropriate for public-facing records, similar to other record types.

jraddaoui commented on js/dialog.js:
```diff
@@ -382,7 +387,15 @@
+                        return { 'results' : [multiColumnResult] };
+                      } else {
+                        // Fallback for pages decorated with 'layout_1col'
+                        return { 'results' : [ $(response).find('h1').text().trim() ] };
```
To avoid unexpected `h1` in some customized headers I'd use `#wrapper h1` as selector. I'd also use `.first()` before getting the text to make sure we have only the first one. And, a suggestion to avoid parsing the response twice into a jQuery var and some (really minor) duplication ...

```javascript
var $response = $(response);

// For pages decorated with 'layout_2col' or 'layout_3col'
var $multiColumnResult = $response.find('#main-column h1');

// Fallback for pages decorated with 'layout_1col'
if (0 == $multiColumnResult.length) {
  $multiColumnResult = $response.find('#wrapper h1');
}

return {'results': [$multiColumnResult.first().text().trim()]};
```

mInnes-archives replied:
I tried using the selector `#wrapper h1` prior to making the commit, but it doesn't seem to want to return anything. It still doesn't, and I have no idea why.

Alternatively, we could just add something like `id='resource-name'` to the `<h1>` we are targeting and use `#resource-name` as the selector.

jraddaoui replied:
Right, I faced that issue a while ago and couldn't find the reason at first glance. Adding an id to the header sounds good @mInnes-archives, it may be needed later one in other pages, but we'll figure out. Thanks!

jraddaoui commented on plugins/sfIsdfPlugin/modules/sfIsdfPlugin/templates/_relatedFunction.php:
```diff
@@ -127,6 +149,12 @@
 
     <div>
 
+     <div class="messages error" id="relatedFunctionError" style="display: none">
```
Nitpick: indentation missing one space.

jraddaoui commented on plugins/sfIsdfPlugin/modules/sfIsdfPlugin/templates/indexSuccess.php:
```diff
@@ -20,6 +17,17 @@
 
+  <section class="breadcrumb">
+    <ul>
+    <li><?php echo link_to(esc_specialchars(sfConfig::get('app_ui_label_function')), ['module' => 'function', 'action' => 'browse']); ?></li>
```
Nitpick: indentation.

jraddaoui commented on plugins/sfIsdfPlugin/modules/sfIsdfPlugin/templates/indexSuccess.php:
```diff
@@ -20,6 +17,17 @@
+<?php end_slot(); ?>
+
+<?php slot('before-content'); ?>
+
```
Apart from helping with the implementation I think this is a good change, but I'd like to check with @fiver-watson ...

Basically, it removes the "View ISDF function" from the function's index page heading and adds a breadcrumb below the function title. This breadcrumb uses the UI label for functions and links to the function's browse page.

fiver-watson replied:
Looks good to me - thanks for noticing and addressing this, @mInnes-archives ! The functions module has been a bit neglected, as it is not frequently used in most North American archival practices (or some European and others - it's most often used in the Australian series system). However, I agree that this is a good change, that brings the page layout more in line with how other entities are presented. Thanks again!

jraddaoui reviewed (approved):
Thanks @mInnes-archives!

I've added some comments below but it looks great to me. I'm not sure if we have a ticket number already and I've pinged @fiver-watson to check the function heading change, but it should be almost ready to squash and merge.

mInnes-archives commented:
Two bugs still need to be addressed.

Mentioned above:
> Dialogs which allow for on-the-fly record creation for new related entities causes both the new record to be created (as desired) as well an additional 'Untitled' record. The new dialog validator also seems to cause some problems when dealing with these newly created entities. Affects the following:
>
> - 'Accession - Related Donors'
> - Actor events for archival descriptions, but note that the 500 error outlined above isn't relevant for this

And:
 
When trying to edit an existing relation (one already in the database), the following series of actions results in the original `null` value of empty fields being changed to a string containing the word `'null'`:

1. click the relation you want edit
2. dialog pops up
3. cancel the dialog
4. save record

Result:
![nullBug](https://user-images.githubusercontent.com/76928773/115744897-f951af00-a360-11eb-978f-e144e10c9002.png)


The 'null' strings can be removed easily by editing the relation again. This appears to affect most relation dialogs.

I can address these over the next little bit.

jraddaoui commented:
Thanks @mInnes-archives.

We have created https://projects.artefactual.com/issues/13499 to reference in the final commit.

mInnes-archives commented:
Since my last comment, the following issues have been addressed:

### Issue 1
(as above)
>When trying to edit an existing relation (one already in the database), the following series of actions results in the original null value of empty fields being changed to a string containing the word 'null':
>
>1. click the relation you want edit
>2. dialog pops up
>3. cancel the dialog
>4. save record

Solution: 0224883993de95101589c389c55bb784c5af8e4e

### Issue 2
(modified from above)
Dialogs with autocomplete fields which allow for the creation of new related resources causes all new records (n>=1) to be created (as desired) as well an additional 'Untitled' record. This appears to be the result of both dialog.js and autocomplete.js trying to add these new related resources, with dialog.js being the only one to do so correctly. I have added an exception in autocomplete.js for the three instances where this behaviour occurs. I couldn't devise a more generalized solution to this issue, so should an autocomplete field with add functionality find its way into a different YUI, this case would need to be added.

Solution: d585b14a285baa04fed8ea6e77287ed076de7762

### Issue 3
Related to Issue 2 above. With both dialog.js and autocomplete.js firing on the same data, the relation trying to be added to the database would sometimes be missing the object, the newly created related resource, resulting in a 500 error from a `NOT NULL` constraint. 

Solution: d585b14a285baa04fed8ea6e77287ed076de7762 (same as Issue 2)

All other commits should be fairly straightforward. I think this covers all of the YUI dialog related bugs that I've found. I'd like a final review and then I'll prepare things for merge.

jraddaoui commented:
Thanks @mInnes-archives, this is looking great and almost ready for merge!

Just one question about the autocomplete exceptions ... do you think it would be feasible to check for the input ancestor elements and stop if there is a `div.yui-dialog`? If that's possible it could prevent having to add new dialog autocomplete inputs in there.

mInnes-archives commented:
> Just one question about the autocomplete exceptions ... do you think it would be feasible to check for the input ancestor elements and stop if there is a `div.yui-dialog`? If that's possible it could prevent having to add new dialog autocomplete inputs in there.

``` javascript
inputAncestor = $input.parent()[0].className;
if (inputAncestor != "form-item yui-ac") {
  $relatedResourceForm = getRelatedResourceForm($input, $hidden, uri, rrFormInputId);
}
```

This seems to work. Are there any issues this might cause elsewhere?

jraddaoui commented:
I'm not sure that will work, `yui-ac` may refer to the autocomplete, so it would be present in all of them. Maybe we could use ...

```js
if ($input.parents('div.yui-dialog').length == 0) {
  $relatedResourceForm = getRelatedResourceForm($input, $hidden, uri, rrFormInputId);
}
```

Or something like that.

mInnes-archives commented:
Okay, I did a bit of testing and found some instances where my solution creates problems. Selecting `yui-dialog` is the way to go.

mInnes-archives commented:
While working on this I found another issue that may be related to dialog/autocomplete. I did a check on the main branch and found the same problem, so it doesn't look like I caused this with anything I've done here.

### Issue 4
When trying to create relationships between the following you receive a 500 error:

- New function [function/add] -> function
- New authority record [actor/add] -> authority record

Establishing these relationships from an existing record [actorSlug/edit; functionSlug/edit] works fine. The fact that these are relationships between the same record types might point to the cause.

I'll look into this since I'd like to fix up all the relation issues in one go, if possible.

mInnes-archives commented:
Works now. New records don't have IDs at this point.

mInnes-archives commented:
From what I can tell, this PR addresses the following open issues:

- Cannot create new authority record and add relationship at the same time
  - https://projects.artefactual.com/issues/10682
  - **[Issue 4 above]**
- Error when adding function related to another
  - https://projects.artefactual.com/issues/13218
  - **[Issue 4 above]**
- Adding multiple event actors on the fly in the RAD template can cause the link to the authority record to be lost on save
  - https://projects.artefactual.com/issues/13425
  - **[Issue 3 above]**

jraddaoui commented on plugins/sfIsaarPlugin/modules/sfIsaarPlugin/actions/relatedAuthorityRecordComponent.class.php:
```diff
@@ -93,7 +93,7 @@ protected function processField($field)
                 if (isset($value)) {
                     $params = $this->context->routing->parse(Qubit::pathInfo($value));
-                    if ($this->resource->id != $this->relation->objectId) {
+                    if ($this->resource->id != $this->relation->objectId || $this->resource->id == null) {
```
Not a big deal, but I think the following is clearer ...

```php
if (!isset($this->resource->id) || $this->resource->id != $this->relation->objectId) {
```

Not in this case, but usually checking the existency before comparing may prevent some issues.

jraddaoui reviewed (approved):
Thanks a lot @mInnes-archives, those are great news!

Added a minor comment below, but feel free to ignore it. Once you clean-up and get this ready for merge, if you're able to reference the three tickets in the commit subject that would be great, just `, refs #12345 #12345 #12345`. If it's a pain to make that fit, just reference the ticket we created for this in the subject and the other two in the body.

mInnes-archives committed:
Relation fixes, refs #10682 #13218 #13425 #13499

- Fixes various 500 errors related to adding relationships
- Adds error messages for required fields in YUI dialogs
- Other relationship related fixes
- Fixes alignment of edit/delete buttons
- Minor title layout change for ISDF index

commit 3c8f88d6351b719870a82121d0d382d290d169c3
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun May 23 13:01:51 2021 -0400

    Rearrange conditional

commit 8c65454f174d8260fa4d1317f0d72077475c22ed
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri May 21 15:48:39 2021 -0400

    Run php-cs-fixer

commit be5eaefd9b12018d8f154ac9424f71968b0f1914
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri May 21 15:46:07 2021 -0400

    Fix ability to add relationship during record creation

commit 6c80b00fa4fc7bf29a8d2aa43d1b8e48c9ed91aa
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue May 18 16:33:15 2021 -0400

    Generalize exception

commit 1110fbcb6cfa2fd13bf1728103fc68f391595167
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue May 11 22:52:11 2021 -0400

    Exclude new related resources to be handled by dialog.js

commit 93a7f43de8a8d23f4e04c3a6f685980b38a2a8fe
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue May 11 22:47:57 2021 -0400

    Typo

commit e84f19caed02133dee467f72be32c9eed23264b1
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri May 7 18:31:29 2021 -0400

    Add Accession related donor required field error

commit 85d77b25666d73c21a2697e354e86ee4bd9c7420
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sat May 1 14:44:55 2021 -0400

    Ensure null values remain null

commit d878f04cf00e3433edebb8fda0fef850b6744826
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri Apr 23 12:23:30 2021 -0400

    Fix indentation

commit 0ff7e0386ad0437c5952df75bc22f1112b0c6302
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri Apr 23 11:23:04 2021 -0400

    Ready required field help for translation

commit 11350dcf5cbdd79f0a476c121633ae19298cde12
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri Apr 23 11:16:17 2021 -0400

    Logic cleanup

commit b54f74e2a3b5cbcd35b544c24fc5b49e4b6cd609
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri Apr 23 11:15:40 2021 -0400

    Logic cleanup

commit 8f3d723faeb49cad41a5456590bf3ba6183c0096
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Fri Apr 23 11:14:08 2021 -0400

    Fix indentation

commit ab970b86c237ed71a4eadd30fa81dc93a190734f
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Apr 21 13:26:29 2021 -0400

    Make title format congruent with other public facing records

commit 00beda9b8aa19ea7b74ee0f1472673f3535f59f7
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Apr 21 13:22:31 2021 -0400

    Accommodate names retrieved from 'layout_1col' pages

commit ccd39c4855724e16651a61594d9e411042e4864f
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Apr 21 13:14:43 2021 -0400

    Add ISDF related function required field error

commit f008cea5fb5ff0b81531ea4015ac3c362754c30b
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:56:16 2021 -0400

    Run php-cs-fixer

commit fe3ef0f2ae874cb7d14dfc4f174bb032a8371d10
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:50:12 2021 -0400

    Fix typo

commit 5a584e4e2fa217b05182bec3b848d3636d32a8b6
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:48:24 2021 -0400

    Make edit button alignment consistent

commit 544b7e06e9936a7070ead8db1630b564b666fa67
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:47:29 2021 -0400

    Add ISDF related resource required field error

commit 51de3504d06f7ef61ee1f83f1f9eac7ba6d806c6
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:46:47 2021 -0400

    Make edit button alignment consistent

commit 506313958f111ab95bfb8446cc75082f002d4f4f
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:45:32 2021 -0400

    Add ISDF related authority record required field error

commit c0b8ac80d29c2bf67c4f342a9abff32145416b21
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:44:57 2021 -0400

    Make edit button alignment consistent

commit 56b49454c833a6b810077fe1ba7f26364a179f89
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:43:55 2021 -0400

    Add ISAAR related authority record required field error

commit c6a5dc78b976eb8c9a22653db48458bcc2b5e5e6
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:42:35 2021 -0400

    Make edit button alignment consistent

commit 2578f21f164c139753a89ebd9fdda12b25d0b53b
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:40:35 2021 -0400

    Add ISAAR related resource required field error

commit 9be9bf2be9ce97c50b311d3538a1ec6c7dcd59b9
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:36:58 2021 -0400

    Add afterCancelLogic option

commit a75f85220d23ded38215a51ebb1de420b62be901
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:35:42 2021 -0400

    Move hide dialog after validation

commit cf0f001fd948a3f4fb19e2ed960fb923a695cd2d
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:22:59 2021 -0400

    Correct field label typo

commit be522787fbadb74e15a9b9b5c7d832e3daaf166b
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 13:21:38 2021 -0400

    Run php-cs-fixer

commit f2140760f8203d9a548c30517c4ee247b0e1e65e
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 12:44:45 2021 -0400

    Revert "Fix indentation"

    This reverts commit a3a5481.

commit abb38407e23f7a14a64af870cead7607f81bb43b
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sun Apr 18 12:44:32 2021 -0400

    Fix indentation

commit 948c08d2e06547b871610b0046f728e838720837
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Mar 31 00:22:25 2021 -0400

    Fix incomplete relationship error

    Fixes 500 error when trying to save incomplete relationships

jraddaoui commented:
Great one @mInnes-archives! Thanks again!

adaxem commented:
Hi all,

Can you please confirm in which version Issue 2 below has been addressed for new Donors created while editing an Accession record? Has the solution been tested in multilingual instances where English is not the main language?

> ### Issue 2
> 
> (modified from above)
> 
> Dialogs with autocomplete fields which allow for the creation of new related resources causes all new records (n>=1) to be created (as desired) as well an additional 'Untitled' record. This appears to be the result of both dialog.js and autocomplete.js trying to add these new related resources, with dialog.js being the only one to do so correctly. I have added an exception in autocomplete.js for the three instances where this behaviour occurs. I couldn't devise a more generalized solution to this issue, so should an autocomplete field with add functionality find its way into a different YUI, this case would need to be added.
> 
> 
> 
> Solution: d585b14a285baa04fed8ea6e77287ed076de7762
>

anvit commented:
Hi @adaxem! These fixes were included in [AtoM v2.7](https://wiki.accesstomemory.org/wiki/Releases/Release_announcements/Release_2.7). As for multilingual instances where English isn't the primary language, we haven't personally tested this fix on one, but definitely let us know if you have any updates or open a new issue if you find out that there is a problem.
